### PR TITLE
Detailed resources

### LINKED ISSUE #1294: Detailed resources reports
We should include sth like `--detailed-resources` flag to print more detailed info about resources (builtins, steps etc.)

### PR DESCRIPTION
**Pull request by drknzz**
Closes #1294 

## Introduced changes

- add `--detailed-resources` option

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: docs/src/appendix/snforge/test.md [hunk 1]
```diff
@@ -52,6 +52,10 @@ Control when colored output is used. Valid values:
 - `always`: always display colors.
 - `never`: never display colors.
 
+## `--detailed-resources`
+
+Display additional info about used resources.
```
Comment (piotmag769) [inline docs/src/appendix/snforge/test.md:57]:
```suggestion
Display additional info about used resources for each test.
```

File: crates/forge/tests/e2e/running.rs [hunk 2]
```diff
@@ -1003,3 +1004,30 @@ fn incompatible_snforge_std_version_warning() {
...
+        Collected 1 test(s) from complex_package package
+        Running 0 test(s) from src/
+        Running 1 test(s) from tests/
+        [PASS] tests::test_contract::test_transfer[..]
+                steps: 6727
+                memory holes: 436
+                builtins: ("range_check_builtin": 215, "pedersen_builtin": 19)
+                syscalls: (CallContract: 5, StorageWrite: 4, GetExecutionInfo: 3, StorageRead: 2, EmitEvent: 1)
+        Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, 0 filtered out
+        [..]
```
Comment (piotmag769) [inline crates/forge/tests/e2e/running.rs:1030]: What is this line matching?

File: crates/forge/tests/e2e/running.rs [hunk 3]
```diff
@@ -1003,3 +1004,30 @@ fn incompatible_snforge_std_version_warning() {
         "}
     );
 }
+
+#[test]
+fn detailed_resources_flag() {
+    let temp = setup_package("complex_package");
```
Comment (piotmag769) [inline crates/forge/tests/e2e/running.rs:1010]: Do we need another cairo package to test that? Cannot we reuse an already existing package?

File: crates/forge-runner/src/printing.rs [hunk 4]
```diff
@@ -37,7 +43,60 @@ pub(crate) fn print_test_result(any_test_result: &AnyTestCaseSummary) {
         }
         _ => String::new(),
     };
-    println!("{result_header} {result_name}{fuzzer_report}{gas_usage}{result_msg}");
+
+    let used_resources = match any_test_result {
+        AnyTestCaseSummary::Single(TestCaseSummary::Passed { used_resources, .. }) => {
+            fn sort_by_value<K, V>(map: &std::collections::HashMap<K, V>) -> Vec<(&K, &V)>
```
Comment (piotmag769) [inline crates/forge-runner/src/printing.rs:49]: Please extract it to separate functions xD

File: crates/forge-runner/src/printing.rs [hunk 5]
```diff
@@ -37,7 +43,60 @@ pub(crate) fn print_test_result(any_test_result: &AnyTestCaseSummary) {
...
+                    .map(|(key, value)| format!("{key:?}: {value}"))
+                    .collect::<Vec<String>>()
+                    .join(", ")
+            }
+
+            let vm_res = &used_resources.execution_resources.vm_resources;
+
+            let sorted_builtins = sort_by_value(&vm_res.builtin_instance_counter);
+            let sorted_syscalls =
+                sort_by_value(&used_resources.execution_resources.syscall_counter);
```
Comment (piotmag769) [inline crates/forge-runner/src/printing.rs:74]: Do we need to sort it? I think printing in the same order consistently is better

File: crates/forge-runner/src/test_case_summary.rs [hunk 6]
```diff
@@ -159,8 +163,9 @@ impl TestCaseSummary<Fuzzing> {
                 TestCaseSummary::Passed {
                     name,
                     msg,
-                    gas_info: GasStatistics::new(&gas_usages),
                     arguments,
+                    gas_info: GasStatistics::new(&gas_usages),
+                    used_resources: UsedResources::default(),
```
Comment (piotmag769) [inline crates/forge-runner/src/test_case_summary.rs:168]: Shouldn't we at least calculate means of used resources?

**Review by piotmag769 — COMMENTED** [review]
I see we don't print used gas for failed tests, what was the design decision for that? I agree that we should be consistent now that we want to print used resources, just curious
cc @Arcticae

File: crates/forge-runner/src/printing.rs [hunk 7]
```diff
@@ -37,7 +43,60 @@ pub(crate) fn print_test_result(any_test_result: &AnyTestCaseSummary) {
         }
         _ => String::new(),
     };
-    println!("{result_header} {result_name}{fuzzer_report}{gas_usage}{result_msg}");
+
+    let used_resources = match any_test_result {
+        AnyTestCaseSummary::Single(TestCaseSummary::Passed { used_resources, .. }) => {
```
Comment (Arcticae) [inline crates/forge-runner/src/printing.rs:48]: Please add an issue to add support for fuzzing as well

File: crates/forge-runner/src/printing.rs [hunk 8]
```diff
@@ -37,7 +43,60 @@ pub(crate) fn print_test_result(any_test_result: &AnyTestCaseSummary) {
...
+    };
+
+    println!(
+        "{result_header} {result_name}{fuzzer_report}{gas_usage}{}{result_msg}",
+        if runner_config.detailed_resources {
+            &used_resources
+        } else {
+            ""
+        }
+    );
```
Comment (Arcticae) [inline crates/forge-runner/src/printing.rs:99]: I feel like this method could be easily made into a builder of some sorts or just refactored to be more legible. Can you submit an issue plz?

File: crates/forge-runner/src/running.rs [hunk 9]
```diff
@@ -187,6 +188,7 @@ fn build_syscall_handler<'a>(
 pub struct RunResultWithInfo {
     pub(crate) run_result: Result<RunResult, RunnerError>,
     pub(crate) gas_used: u128,
+    pub(crate) used_resources: UsedResources,
```
Comment (Arcticae) [inline crates/forge-runner/src/running.rs:191]: I would advise to merge those 2 into just one `resources`

File: crates/forge-runner/src/test_case_summary.rs (hunk 6 above)
Comment (Arcticae) [reply crates/forge-runner/src/test_case_summary.rs:168]: Let's do it in another PR

File: crates/forge/tests/e2e/running.rs (hunk 3 above)
Comment (Arcticae) [reply crates/forge/tests/e2e/running.rs:1010]: Agree on that with Piotr

**Review by Arcticae — CHANGES_REQUESTED** [review]
I think we also need some simple tests (integration ones) in terms of the correctness of the resources returned. Please refer to gas tests in order to see what i mean - those need to be simple and verifiable "by hand" if they are correct.

File: crates/forge/tests/e2e/running.rs [hunk 10]
```diff
@@ -1003,3 +1003,29 @@ fn incompatible_snforge_std_version_warning() {
...
+        Collected 1 test(s) from erc20_package package
+        Running 0 test(s) from src/
+        Running 1 test(s) from tests/
+        [PASS] tests::test_complex::complex[..]
+                steps: 5621
+                memory holes: 330
+                builtins: ("range_check_builtin": 186, "pedersen_builtin": 23)
+                syscalls: (StorageRead: 22, StorageWrite: 12, EmitEvent: 4, GetExecutionInfo: 3)
+        Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, 0 filtered out
+        "#}
```
[the remark this answers is not in the document]
Comment (Arcticae) [reply crates/forge/tests/e2e/running.rs:1029]: This can be done in scope of another issue/PR

File: crates/forge-runner/src/printing.rs [hunk 11]
```diff
@@ -37,7 +43,59 @@ pub(crate) fn print_test_result(any_test_result: &AnyTestCaseSummary) {
         }
         _ => String::new(),
     };
-    println!("{result_header} {result_name}{fuzzer_report}{gas_usage}{result_msg}");
+
+    let used_resources = match any_test_result {
+        AnyTestCaseSummary::Single(TestCaseSummary::Passed { used_resources, .. }) => {
+            let vm_res = &used_resources.execution_resources.vm_resources;
```
Comment (war-in) [inline crates/forge-runner/src/printing.rs:49]: we should not prepare resources if the flag is not passed

File: crates/forge/tests/data/erc20_package/tests/test_complex.cairo [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    let mut initial_supply_serialized = array![];
+    initial_supply.serialize(ref initial_supply_serialized);
+
+    constructor_calldata.append_span(initial_supply_serialized.span());
+    constructor_calldata.append(recipient.into());
+
+    contract.deploy(@constructor_calldata).unwrap()
+}
+
+// Syscalls from constructor are not counted
```
Comment (war-in) [inline crates/forge/tests/data/erc20_package/tests/test_complex.cairo:30]: Is this a bug or a feature? 🤔

File: crates/forge/tests/e2e/running.rs [hunk 13]
```diff
@@ -1003,3 +1003,29 @@ fn incompatible_snforge_std_version_warning() {
...
+        
+
+        Collected 1 test(s) from erc20_package package
+        Running 0 test(s) from src/
+        Running 1 test(s) from tests/
+        [PASS] tests::test_complex::complex[..]
+                steps: 5621
+                memory holes: 330
+                builtins: ("range_check_builtin": 186, "pedersen_builtin": 23)
+                syscalls: (StorageRead: 22, StorageWrite: 12, EmitEvent: 4, GetExecutionInfo: 3)
```
Comment (war-in) [inline crates/forge/tests/e2e/running.rs:1027]: Would be nice to have a `\n` here

File: crates/forge/tests/e2e/running.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (war-in) [reply crates/forge/tests/e2e/running.rs:1029]: What do you mean by `our test env`?

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/rpc.rs [hunk 14]
```diff
@@ -31,9 +31,19 @@ pub struct UsedResources {
 impl UsedResources {
     pub fn extend(self: &mut UsedResources, other: &UsedResources) {
         self.execution_resources.vm_resources += &other.execution_resources.vm_resources;
-        self.execution_resources
+
+        other
+            .execution_resources
             .syscall_counter
-            .extend(&other.execution_resources.syscall_counter);
```
Comment (drknzz) [inline crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/rpc.rs:36]: Funny, this caused syscalls to not accumulate, as `map1.extend(map2)` replaced already existing entries in `map1` with entries in `map2`

File: crates/forge-runner/src/printing.rs [hunk 15]
```diff
@@ -37,7 +43,59 @@ pub(crate) fn print_test_result(any_test_result: &AnyTestCaseSummary) {
...
+}
+
+fn format_items<K, V>(items: &[(K, V)]) -> String
+where
+    K: std::fmt::Debug,
+    V: std::fmt::Display,
+{
+    items
+        .iter()
+        .map(|(key, value)| format!("{key:?}: {value}"))
```
[the remark this answers is not in the document]
Comment (drknzz) [reply crates/forge-runner/src/printing.rs:95]: `DeprecatedSyscallSelector` doesn't implement `Display` :/

File: crates/forge-runner/src/running.rs (hunk 9 above)
Comment (drknzz) [reply crates/forge-runner/src/running.rs:191]: Can't really put `gas_used` inside of `UsedResources`, as gas is calculated later, and it uses `UsedResources`. There could be a new struct `Resources` these 2, but I don't know if it wouldn't just introduce complexity.

File: crates/forge/tests/data/erc20_package/tests/test_complex.cairo (hunk 12 above)
Comment (drknzz) [reply crates/forge/tests/data/erc20_package/tests/test_complex.cairo:30]: That's the part I'm not sure about 🧐

File: crates/forge/tests/e2e/running.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Arcticae) [reply crates/forge/tests/e2e/running.rs]: I mean our node (testnet/integr)

File: crates/forge-runner/src/running.rs (hunk 9 above)
Comment (Arcticae) [reply crates/forge-runner/src/running.rs:168]: If it's not feasible i'm not going to push it

File: crates/forge-runner/src/test_case_summary.rs (hunk 6 above)
Comment (Arcticae) [reply crates/forge-runner/src/test_case_summary.rs:179]: @drknzz would you submit an issue?

Comment (drknzz) [reply crates/forge-runner/src/test_case_summary.rs:179]: Yes, will submit all of the further issues from this pr before merge

File: crates/forge/test_utils/src/runner.rs [hunk 16]
```diff
@@ -295,7 +295,86 @@ macro_rules! assert_gas {
                     panic!("Cannot use assert_gas! for fuzzing tests")
                 }
                 AnyTestCaseSummary::Single(case) => match case {
-                    TestCaseSummary::Passed { gas_info: gas, .. } => *gas == $asserted_gas,
+                    TestCaseSummary::Passed { gas_info: gas, .. } => {
+                        *gas == $asserted_gas
+                            && any_case
+                                .name()
+                                .unwrap()
+                                .ends_with(test_name_suffix.as_str())
```
Comment (Arcticae) [inline crates/forge/test_utils/src/runner.rs:303]: Why is this needed? Isn't it filtered already by now?

File: crates/forge/tests/e2e/running.rs [hunk 17]
```diff
@@ -997,3 +997,30 @@ fn incompatible_snforge_std_version_warning() {
         "}
     );
 }
+
+#[test]
+fn detailed_resources_flag() {
+    let temp = setup_package("erc20_package");
```
Comment (Arcticae) [inline crates/forge/tests/e2e/running.rs:1003]: I would get rid of this test, this will just make us update mindlessly those numbers when something changes and it doesn't really change anything other than stdout format

File: crates/forge/tests/integration/resources.rs [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+    );
+
+    let result = run_test_case(&test);
+
+    assert_passed!(result);
+
+    assert_syscalls!(result, "keccak", Keccak, 1);
+    assert_syscalls!(result, "deploy", Deploy, 1);
+    assert_syscalls!(result, "storage_read", StorageRead, 1);
+    assert_syscalls!(result, "storage_write", StorageRead, 1);
```
Comment (Arcticae) [inline crates/forge/tests/integration/resources.rs:143]:
```suggestion
    assert_syscalls!(result, "storage_write", StorageWrite, 1);
```

File: crates/forge/tests/integration/resources.rs [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+                assert(1 == 1, 'error message');
+            }
+        "
+    ));
+
+    let result = run_test_case(&test);
+
+    assert_passed!(result);
+
+    // No ECDSA and Keccak builtins
```
Comment (Arcticae) [inline crates/forge/tests/integration/resources.rs:51]: Why not? 🤔

File: crates/forge/tests/e2e/running.rs (hunk 17 above)
Comment (drknzz) [reply crates/forge/tests/e2e/running.rs:1003]: The numbers could be omitted
```                
steps: [..]
memory holes: [..]
builtins: ([..])
syscalls: ([..])
```

File: crates/forge/test_utils/src/runner.rs (hunk 16 above)
Comment (drknzz) [reply crates/forge/test_utils/src/runner.rs:303]: Apparently, it's not. It just worked because there was always a single cairo test in the code.

File: crates/forge/tests/integration/resources.rs (hunk 19 above)
Comment (drknzz) [reply crates/forge/tests/integration/resources.rs:51]: I haven't found (and neither did @war-in when implementing gas) any way to make these builtins execute. Keccak is counted as a syscall in the test below though.

File: crates/forge/tests/e2e/running.rs (hunk 17 above)
Comment (Arcticae) [reply crates/forge/tests/e2e/running.rs:1003]: please do that if you'd like to keep that one

File: crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/rpc.rs [hunk 20]
```diff
@@ -31,12 +32,22 @@ pub struct UsedResources {
...
+
+    fn update_syscall_counter(self: &mut UsedResources, syscall_counter: &SyscallCounter) {
+        for (syscall, count) in syscall_counter {
+            *self
+                .execution_resources
+                .syscall_counter
+                .entry(*syscall)
+                .or_insert(0) += count;
+        }
+    }
```
Comment (Arcticae) [inline crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/rpc.rs:50]: Also i would add a test for that since our earlier tests didn't catch that

Comment (drknzz) [reply crates/cheatnet/src/runtime_extensions/call_to_blockifier_runtime_extension/rpc.rs:50]: Yup, already added: [Add failing syscalls test](https://github.com/foundry-rs/starknet-foundry/pull/1511/commits/0a30c1eda077f84344610c0fc591d9a89d8a236f)

File: crates/forge/tests/e2e/running.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/tests/e2e/running.rs]: https://github.com/starkware-libs/starknet-specs/blob/49665932a97f8fdef7ac5869755d2858c5e3a687/api/starknet_trace_api_openrpc.json#L11
This should work

[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/tests/e2e/running.rs]: Example call
```console
$ curl -XPOST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","method":"starknet_traceTransaction","id":0,"params":{"transaction_hash": "0x5e2397ed1acc530208567412ed3b1959a174e42c2b2a739c44ab063d7edcc1f"}}' "http://188.34.188.184:9545/rpc/v0_6"
```

File: crates/forge/tests/data/erc20_package/tests/test_complex.cairo (hunk 12 above)
Comment (piotmag769) [reply crates/forge/tests/data/erc20_package/tests/test_complex.cairo:30]: We need an issue for that

File: crates/forge/src/main.rs [hunk 21]
```diff
@@ -91,6 +91,10 @@ struct TestArgs {
     #[arg(long, conflicts_with = "only_ignored")]
     include_ignored: bool,
 
+    /// Display more detailed info about used resources
+    #[arg(long)]
+    detailed_resources: bool,
```
Comment (piotmag769) [inline crates/forge/src/main.rs:96]: We could add that configuration option to Scarb.toml, seems tedious to always write it by hand

File: docs/src/appendix/snforge/test.md [hunk 22]
```diff
@@ -52,9 +52,13 @@ Control when colored output is used. Valid values:
 - `always`: always display colors.
 - `never`: never display colors.
 
+## `--detailed-resources`
+
+Display additional info about used resources for each test.
```
Comment (piotmag769) [inline docs/src/appendix/snforge/test.md:57]:
```suggestion
Display additional info about used resources for passed tests.
```

File: crates/forge/tests/integration/gas.rs [hunk 23]
```diff
@@ -532,9 +532,9 @@ fn multiple_storage_writes_cost() {
     let result = run_test_case(&test);
 
     assert_passed!(result);
-    // 3 = gas cost of steps
+    // 4 = gas cost of steps
```
Comment (piotmag769) [inline crates/forge/tests/integration/gas.rs:535]: Are you sure? Nothing else changed and steps were calculated correctly, prob another builtin (that was previously calculated incorrectly) is responsible for this (see [this](https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/fee-mechanism/#gas_cost_per_cairo_step_or_builtin_step) and the formula above)

File: crates/forge/tests/e2e/running.rs (hunk 17 above)
Comment (piotmag769) [reply crates/forge/tests/e2e/running.rs:1003]: We need to keep this one imo - it checks the e2e functionality (if we print the resources at all). Agree with omitting the numbers though

File: crates/forge/test_utils/src/runner.rs [hunk 24]
```diff
@@ -295,7 +295,86 @@ macro_rules! assert_gas {
...
+                    }
+                    _ => false,
+                },
+            }
+        }));
+    }};
+}
+
+#[macro_export]
+macro_rules! assert_syscalls {
```
Comment (piotmag769) [inline crates/forge/test_utils/src/runner.rs:313]:
```suggestion
macro_rules! assert_syscall {
```
It asserts one syscall only

File: crates/forge/test_utils/src/runner.rs [hunk 25]
```diff
@@ -295,7 +295,86 @@ macro_rules! assert_gas {
...
+                    }
+                    _ => false,
+                },
+            }
+        }));
+    }};
+}
+
+#[macro_export]
+macro_rules! assert_builtins {
```
Comment (piotmag769) [inline crates/forge/test_utils/src/runner.rs:349]:
```suggestion
macro_rules! assert_builtin {
```
It asserts one builtin only

**Review by piotmag769 — APPROVED** [review]
LGTM, please address the comments

File: crates/forge/tests/integration/gas.rs (hunk 23 above)
Comment (drknzz) [reply crates/forge/tests/integration/gas.rs:535]: The cost has increased, because now the syscalls is counted twice, instead of once [here](https://github.com/starkware-libs/blockifier/blob/8f423a2f5c351e4c22c7350b08afd8082fb0fa9f/crates/blockifier/src/fee/os_usage.rs#L46).

File: crates/forge-runner/src/test_case_summary.rs (hunk 6 above)
Comment (drknzz) [reply crates/forge-runner/src/test_case_summary.rs:179]: #1628

File: crates/forge-runner/src/printing.rs (hunk 7 above)
Comment (drknzz) [reply crates/forge-runner/src/printing.rs]: #1628

File: crates/forge/tests/e2e/running.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (drknzz) [reply crates/forge/tests/e2e/running.rs]: #1630

File: crates/forge/tests/data/erc20_package/tests/test_complex.cairo (hunk 12 above)
Comment (drknzz) [reply crates/forge/tests/data/erc20_package/tests/test_complex.cairo:30]: #1631

File: docs/src/testing/testing.md [hunk 26]
```diff
@@ -124,3 +124,20 @@ Tests: 0 passed, 0 failed, 0 skipped, 1 ignored, 0 filtered out
 
 To run only tests marked with the  `#[ignore]` attribute use `snforge test --ignored`. 
 To run all tests regardless of the `#[ignore]` attribute use `snforge test --include-ignored`.
+
+## Displaying Resources Used During Test
```
Comment (piotmag769) [inline docs/src/testing/testing.md:128]:
```suggestion
## Displaying Resources Used During Tests
```

File: crates/forge/tests/integration/gas.rs (hunk 23 above)
Comment (piotmag769) [reply crates/forge/tests/integration/gas.rs:535]: True, that means that #1631 is higher priority than I thought

**Review by piotmag769 — APPROVED** [review]
